{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bc6a1419-4eb4-412a-9339-ae6dbe23db19",
   "metadata": {},
   "source": [
    "# Haplotype generator\n",
    "\n",
    "This is to generate haplotypes\n",
    "\n",
    "Input:\n",
    "   - A bed file containing Chromorom   Start   End    Ref   Alt  reg_chrom  reg_start reg_end \n",
    "   - A matching reference genome in FASTA format (hg19, hg38 etc.) \n",
    "   - Note: \n",
    "       - Alt allele can only be one kind, for multiple alleles, create a different input file. \n",
    "       - '-' or '*' means deletion\n",
    "\n",
    "Output: \n",
    "   - FASTA file containing modifications specified in the input. Each sequence is extracted using reg_chrom reg_start reg_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "3f1564be-ae6c-4ff1-9f16-e7a3cf5cffdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os \n",
    "\n",
    "def load_ref(fastafile, target_chrom=None):\n",
    "    \"\"\" \n",
    "    Load the reference genome from a single fasta file. \n",
    "    - If sequence name is specified, only load that sequence.\n",
    "    \"\"\"\n",
    "    chroms = open(fastafile).read().split(\">\")[1:]\n",
    "    genome = {}\n",
    "    for chrom in chroms:\n",
    "        tmp = chrom.split()\n",
    "        chr_name = tmp[0]\n",
    "        if target_chrom != None:\n",
    "            if chr_name != target_chrom:\n",
    "                continue\n",
    "        seq = ''.join(tmp[1:])\n",
    "        genome[chr_name] = seq\n",
    "    return genome\n",
    "\n",
    "def get_roi_seqs(bedFile, refFasta):\n",
    "    \"\"\"\n",
    "    1. From region of interest BED file, output FASTA file containig\n",
    "    sequences for each region.\n",
    "    2. Load sequences for each region\n",
    "    \"\"\"\n",
    "    tmpFAFile = bedFile + '.tmpfasta'\n",
    "    subprocess.run(['bedtools', 'getfasta', '-fi', refFasta, '-bed', bedFile, '-fo', tmpFAFile])\n",
    "    refSeqs = load_ref(tmpFAFile)\n",
    "    os.remove(tmpFAFile)\n",
    "\n",
    "    return refSeqs\n",
    "\n",
    "def convert_roi_names_to_coordinates(listOfROINames):\n",
    "    listOfROIs = []\n",
    "    for roi in listOfROINames:\n",
    "        tmp = roi.split(\":\")\n",
    "        tmp = [tmp[0]] +  [int(x) for x in tmp[1].split('-')]\n",
    "        listOfROIs += [tmp]\n",
    "    return listOfROIs\n",
    "\n",
    "def assign_to_roi(loc, listOfROIs, one_based=False):\n",
    "    \"\"\"\n",
    "    From a location, find a region of interest that contain said location and output new loc.\n",
    "    INPUT:\n",
    "        - location in the form of [CHROMOSOME, POS] # note, POS is 0-based\n",
    "        - ROIs in the form of [[CHROMOSOME, START, END]] # START and END are also 0-based\n",
    "    OUTPUT:\n",
    "        - ROI that contains the input location\n",
    "        - New cordinates with respect to that ROI in the form of [ROI, START, END]\n",
    "    \"\"\"\n",
    "    for roi in listOfROIs:\n",
    "        if roi[0] == loc[0]:\n",
    "            if one_based:\n",
    "                loc[1] = loc[1] - 1 \n",
    "            if roi[1] <= loc[1] and loc[1] < roi[2]:\n",
    "                newPos = loc[1] - roi[1]\n",
    "                roiName = roi[0] + ':' + str(roi[1]) + '-' + str(roi[2])\n",
    "                return roiName, newPos\n",
    "    return None\n",
    "\n",
    "def load_hap_gen_input(hap_gen_bed):\n",
    "    \"\"\"\n",
    "    Read haplotype generator input BED file\n",
    "    TODO: add function to assess input file type\n",
    "    \"\"\"\n",
    "    hap_gen_input = []\n",
    "    for line in open(hap_gen_bed, 'r'):\n",
    "        line = line.strip().split('\\t')\n",
    "        if len(line) > 3:\n",
    "            line[1] = int(line[1])\n",
    "            hap_gen_input += [line]\n",
    "    return hap_gen_input\n",
    "\n",
    "def convert_hapGenInput_to_new_coords(hapGenInput, listOfROIs, one_based=False):\n",
    "    \"\"\"\n",
    "    Convert the hapGenInput to new coordinates based on the reference sequences\n",
    "    \n",
    "    If the locations of variants are 1-based, use one_based = True.\n",
    "    \"\"\"\n",
    "    newHapGenInput = []\n",
    "    for line in hapGenInput:\n",
    "        loc = line[:2]\n",
    "        try:\n",
    "            roi, newPos = assign_to_roi(loc, listOfROIs, one_based=one_based)\n",
    "        except TypeError: # the return is None, the variant is not inside any ROI\n",
    "            continue\n",
    "        newHapGenInput += [[roi, newPos] + line[2:]]\n",
    "        \n",
    "    return newHapGenInput\n",
    "\n",
    "def insert_variants(ref_seq, seq_name, haplotype_calls, one_based=False): \n",
    "    # ref_seq is the seqeuence of the chrom of interest\n",
    "    # seq_name is the name of the ref_seq\n",
    "    # haplotype_calls is in the form ['chr1', '30001230', 'G', 'A']\n",
    "\n",
    "    # copy the seqs first, also convert the string into list of characters\n",
    "    new_seq = ''\n",
    "    counter = 0\n",
    "    new_coords = [] # this stores the coordinates of the inserted variants on the newly altered chromosome\n",
    "    old_coords = [] # this stores old coordinates of inserted variants\n",
    "    prev_var_coord = 0 # tracks the location of the end of the previous variant (VCF coordinate is 1-based), this is on the reference genome\n",
    "    total_added = 0 # tracks the difference in bp number after altering chrom\n",
    "    for line in haplotype_calls: # $haplotypes is in the form [[chrom, loc, ref, alts, gt]]\n",
    "        #print(line)\n",
    "        chrom = line[0]\n",
    "        if chrom != seq_name:\n",
    "            continue\n",
    "        counter += 1\n",
    "        if counter % 10000 == 0:\n",
    "            print(counter)\n",
    "\n",
    "        if one_based:\n",
    "            loc_s = int(line[1]) - 1 # convert 1-based to 0-based \n",
    "        else:\n",
    "            loc_s = int(line[1])\n",
    "\n",
    "        ref = line[2]\n",
    "        loc_e = loc_s + len(ref) # loc_s and loc_e are coords on the reference genome\n",
    "        alt = line[3].replace(\"-\",'').replace(\"*\",'') # if '-', it's a deletion\n",
    "        new_seq += ref_seq[prev_var_coord:loc_s] # adding the space between the current variant and the previous one\n",
    "        new_coords += [len(new_seq)]\n",
    "        old_coords += [loc_s]\n",
    "        new_seq = new_seq + alt\n",
    "        prev_var_coord = loc_e\n",
    "        total_added += len(alt) - len(ref)\n",
    "    new_seq += ref_seq[prev_var_coord:]\n",
    "\n",
    "    #print(\"Inserted number of variants\", counter)\n",
    "    #print(\"Total bases dif\", total_added)\n",
    "\n",
    "    return new_seq, old_coords, new_coords\n",
    "\n",
    "def generate_haplotype_seqs(hapGenInput, refSeqs, one_based=False, tag=''):\n",
    "    \"\"\"\n",
    "    INPUT: Haplotype Input, reference sequences\n",
    "    OUTPUT: Haplotype sequences in FASTA format\n",
    "    \"\"\"\n",
    "    haploSeqs = {}\n",
    "    for seq in refSeqs:\n",
    "        newSeq, old_coords, new_coords = insert_variants(refSeqs[seq], seq, hapGenInput, one_based=one_based)\n",
    "        haploSeqs[seq + '_' + tag] = newSeq\n",
    "    return haploSeqs\n",
    "\n",
    "def main():\n",
    "    \"\"\"\n",
    "    This script is just to create haplotypes based on the requirements \n",
    "    \"\"\"\n",
    "    \n",
    "    return \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "3bfb2e81-495d-4866-a24e-c17638b40761",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr1:115248070-115248190_hap1': 'TATGACAGCTGGCCAGGCAVuTTAATTTGGGGAAAGAGAAGGATTTTGAGGTAAACTAGAATTCTTTTCAGCAATCACAAACATAGCCTAAAAACCACGTTTTATACCACACTGAAGTTA',\n",
       " 'chr1:115251154-115251274_hap1': 'GTvuvuvuvuACATCACCACACATGGCAATCCCATACAACCCTGAGTCCCATCATCACTGCTGTTGAGTTTTTTCATTCGGTACTGGCGTATTTCTCTTACCAGTGTGTAAAAAGCATCTTCAACAC'}"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bedFile = \"/home/vngo/Work/liquid_biopsy_pipeline/mnv_calling_module/target_beds/somaticCNV_from_SomaticGermline_panel.bed\"\n",
    "refFasta = \"/home/vngo/genomes/hg19_ambry.fa\"\n",
    "\n",
    "refSeqs = get_roi_seqs(bedFile, refFasta)\n",
    "roiNames = ['chr1:115248070-115248190', 'chr1:115251154-115251274']\n",
    "# reduce the refSeqs to smaller set for testing\n",
    "tmp = {}\n",
    "for roi in roiNames:\n",
    "    tmp[roi] = refSeqs[roi]\n",
    "refSeqs = tmp\n",
    "\n",
    "listOfROIs = convert_roi_name_to_coordinates(roiNames)\n",
    "\n",
    "#for roi in roiNames:\n",
    "#    print(roi, refSeqs[roi])\n",
    "\n",
    "hap_gen_bed = \"hap_gen_input.bed\"\n",
    "hapGenInput = load_hap_gen_input(hap_gen_bed)\n",
    "newHapGenInput = convert_hapGenInput_to_new_coords(hapGenInput, listOfROIs, one_based=False)\n",
    "newHapGenInput\n",
    "\n",
    "\n",
    "haploSeqs = generate_haplotype_seqs(newHapGenInput, refSeqs, one_based=False, tag='hap1')\n",
    "haploSeqs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "1cbe65b9-9959-4255-bb15-df5459e96f23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test: b''\n",
      "test: b'    ====================ART===================='\n",
      "test: b'             ART_Illumina (2008-2016)'\n",
      "test: b'          Q Version 2.5.8 (June 7, 2016)'\n",
      "test: b'     Contact: Weichun Huang <whduke@gmail.com>'\n",
      "test: b'    -------------------------------------------'\n",
      "test: b''\n",
      "test: b'===== USAGE ====='\n",
      "test: b''\n",
      "test: b'art_illumina [options] -ss <sequencing_system> -sam -i <seq_ref_file> -l <read_length> -f <fold_coverage> -o <outfile_prefix>'\n",
      "test: b'art_illumina [options] -ss <sequencing_system> -sam -i <seq_ref_file> -l <read_length> -c <num_reads_per_sequence> -o <outfile_prefix>'\n",
      "test: b'art_illumina [options] -ss <sequencing_system> -sam -i <seq_ref_file> -l <read_length> -f <fold_coverage> -m <mean_fragsize> -s <std_fragsize> -o <outfile_prefix>'\n",
      "test: b'art_illumina [options] -ss <sequencing_system> -sam -i <seq_ref_file> -l <read_length> -c <num_reads_per_sequence> -m <mean_fragsize> -s <std_fragsize> -o <outfile_prefix>'\n",
      "test: b''\n",
      "test: b'===== PARAMETERS ====='\n",
      "test: b''\n",
      "test: b'  -1   --qprof1   the first-read quality profile'\n",
      "test: b'  -2   --qprof2   the second-read quality profile'\n",
      "test: b'  -amp --amplicon amplicon sequencing simulation'\n",
      "test: b'  -c   --rcount   number of reads/read pairs to be generated per sequence/amplicon (not be used together with -f/--fcov)'\n",
      "test: b'  -d   --id       the prefix identification tag for read ID'\n",
      "test: b'  -ef  --errfree  indicate to generate the zero sequencing errors SAM file as well the regular one'\n",
      "test: b'                  NOTE: the reads in the zero-error SAM file have the same alignment positions'\n",
      "test: b'                  as those in the regular SAM file, but have no sequencing errors'\n",
      "test: b'  -f   --fcov     the fold of read coverage to be simulated or number of reads/read pairs generated for each amplicon'\n",
      "test: b'  -h   --help     print out usage information'\n",
      "test: b'  -i   --in       the filename of input DNA/RNA reference'\n",
      "test: b'  -ir  --insRate  the first-read insertion rate (default: 0.00009)'\n",
      "test: b'  -ir2 --insRate2 the second-read insertion rate (default: 0.00015)'\n",
      "test: b'  -dr  --delRate  the first-read deletion rate (default:  0.00011)'\n",
      "test: b'  -dr2 --delRate2 the second-read deletion rate (default: 0.00023)'\n",
      "test: b'  -k   --maxIndel the maximum total number of insertion and deletion per read (default: up to read length)'\n",
      "test: b'  -l   --len      the length of reads to be simulated'\n",
      "test: b'  -m   --mflen    the mean size of DNA/RNA fragments for paired-end simulations'\n",
      "test: b'  -mp  --matepair indicate a mate-pair read simulation'\n",
      "test: b\"  -M  --cigarM    indicate to use CIGAR 'M' instead of '=/X' for alignment match/mismatch\"\n",
      "test: b\"  -nf  --maskN    the cutoff frequency of 'N' in a window size of the read length for masking genomic regions\"\n",
      "test: b\"                  NOTE: default: '-nf 1' to mask all regions with 'N'. Use '-nf 0' to turn off masking\"\n",
      "test: b'  -na  --noALN    do not output ALN alignment file'\n",
      "test: b'  -o   --out      the prefix of output filename'\n",
      "test: b'  -p   --paired   indicate a paired-end read simulation or to generate reads from both ends of amplicons'\n",
      "test: b'                  NOTE: art will automatically switch to a mate-pair simulation if the given mean fragment size >= 2000'\n",
      "test: b'  -q   --quiet    turn off end of run summary'\n",
      "test: b'  -qL  --minQ     the minimum base quality score'\n",
      "test: b'  -qU  --maxQ     the maxiumum base quality score'\n",
      "test: b'  -qs  --qShift   the amount to shift every first-read quality score by'\n",
      "test: b'  -qs2 --qShift2  the amount to shift every second-read quality score by'\n",
      "test: b'                  NOTE: For -qs/-qs2 option, a positive number will shift up quality scores (the max is 93)'\n",
      "test: b'                  that reduce substitution sequencing errors and a negative number will shift down'\n",
      "test: b'                  quality scores that increase sequencing errors. If shifting scores by x, the error'\n",
      "test: b'                  rate will be 1/(10^(x/10)) of the default profile.'\n",
      "test: b'  -rs  --rndSeed  the seed for random number generator (default: system time in second)'\n",
      "test: b'                  NOTE: using a fixed seed to generate two identical datasets from different runs'\n",
      "test: b'  -s   --sdev     the standard deviation of DNA/RNA fragment size for paired-end simulations.'\n",
      "test: b'  -sam --samout   indicate to generate SAM alignment file'\n",
      "test: b'  -sp  --sepProf  indicate to use separate quality profiles for different bases (ATGC)'\n",
      "test: b'  -ss  --seqSys   The name of Illumina sequencing system of the built-in profile used for simulation'\n",
      "test: b'       NOTE: sequencing system ID names are:'\n",
      "test: b'            GA1 - GenomeAnalyzer I (36bp,44bp), GA2 - GenomeAnalyzer II (50bp, 75bp)'\n",
      "test: b'           HS10 - HiSeq 1000 (100bp),          HS20 - HiSeq 2000 (100bp),      HS25 - HiSeq 2500 (125bp, 150bp)'\n",
      "test: b'           HSXn - HiSeqX PCR free (150bp),     HSXt - HiSeqX TruSeq (150bp),   MinS - MiniSeq TruSeq (50bp)'\n",
      "test: b'           MSv1 - MiSeq v1 (250bp),            MSv3 - MiSeq v3 (250bp),        NS50 - NextSeq500 v2 (75bp)'\n",
      "test: b'===== NOTES ====='\n",
      "test: b''\n",
      "test: b'* ART by default selects a built-in quality score profile according to the read length specified for the run.'\n",
      "test: b''\n",
      "test: b'* For single-end simulation, ART requires input sequence file, output file prefix, read length, and read count/fold coverage.'\n",
      "test: b''\n",
      "test: b'* For paired-end simulation (except for amplicon sequencing), ART also requires the parameter values of'\n",
      "test: b'  the mean and standard deviation of DNA/RNA fragment lengths'\n",
      "test: b''\n",
      "test: b'===== EXAMPLES ====='\n",
      "test: b''\n",
      "test: b' 1) single-end read simulation'\n",
      "test: b' \\tart_illumina -ss HS25 -sam -i reference.fa -l 150 -f 10 -o single_dat'\n",
      "test: b''\n",
      "test: b' 2) paired-end read simulation'\n",
      "test: b'       art_illumina -ss HS25 -sam -i reference.fa -p -l 150 -f 20 -m 200 -s 10 -o paired_dat'\n",
      "test: b''\n",
      "test: b' 3) mate-pair read simulation'\n",
      "test: b'       art_illumina -ss HS10 -sam -i reference.fa -mp -l 100 -f 20 -m 2500 -s 50 -o matepair_dat'\n",
      "test: b''\n",
      "test: b\" 4) amplicon sequencing simulation with 5' end single-end reads\"\n",
      "test: b' \\tart_illumina -ss GA2 -amp -sam -na -i amp_reference.fa -l 50 -f 10 -o amplicon_5end_dat'\n",
      "test: b''\n",
      "test: b' 5) amplicon sequencing simulation with paired-end reads'\n",
      "test: b'       art_illumina -ss GA2 -amp -p -sam -na -i amp_reference.fa -l 50 -f 10 -o amplicon_pair_dat'\n",
      "test: b''\n",
      "test: b' 6) amplicon sequencing simulation with matepair reads'\n",
      "test: b'       art_illumina -ss MSv1 -amp -mp -sam -na -i amp_reference.fa -l 150 -f 10 -o amplicon_mate_dat'\n",
      "test: b''\n",
      "test: b' 7) generate an extra SAM file with zero-sequencing errors for a paired-end read simulation'\n",
      "test: b'       art_illumina -ss HSXn -ef -i reference.fa -p -l 150 -f 20 -m 200 -s 10 -o paired_twosam_dat'\n",
      "test: b''\n",
      "test: b' 8) reduce the substitution error rate to one 10th of the default profile'\n",
      "test: b'       art_illumina -i reference.fa -qs 10 -qs2 10 -l 50 -f 10 -p -m 500 -s 10 -sam -o reduce_error'\n",
      "test: b''\n",
      "test: b\" 9) turn off the masking of genomic regions with unknown nucleotides 'N'\"\n",
      "test: b'       art_illumina -ss HS20 -nf 0  -sam -i reference.fa -p -l 100 -f 20 -m 200 -s 10 -o paired_nomask'\n",
      "test: b''\n",
      "test: b\" 10) masking genomic regions with >=5 'N's within the read length 50\"\n",
      "test: b'       art_illumina -ss HSXt -nf 5 -sam -i reference.fa -p -l 150 -f 20 -m 200 -s 10 -o paired_maskN5'\n",
      "test: b''\n"
     ]
    }
   ],
   "source": [
    "ART=\"/mnt/vngo/art_bin_MountRainier/art_illumina\"\n",
    "\n",
    "proc = subprocess.Popen([ART], stdout=subprocess.PIPE)\n",
    "while True:\n",
    "    line = proc.stdout.readline()\n",
    "    if not line:\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2466f151-f9e2-4f78-a361-4a367ca024d9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base_kernel",
   "language": "python",
   "name": "base_kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
